File: BinaryNotesMQ/.net/BinaryNotesMQ/src/org/bn/mq/net/tcp/TransportFactory.cs
/*
* Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
* 
* Licensed under the LGPL, Version 2 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*      http://www.gnu.org/copyleft/lgpl.html
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* 
* With any your questions welcome to my e-mail 
* or blog at http://abdulla-a.blogspot.com.
*/
using System;
using System.Threading;
using System.Collections.Generic;
using org.bn.mq.protocol;
using org.bn.mq.net;

namespace org.bn.mq.net.tcp
{
	
	public class TransportFactory : ITransportFactory
	{
		virtual public ITransportMessageCoderFactory TransportMessageCoderFactory
		{
			get
			{
				return messageCoderFactory;
			}
			
			set
			{
				this.messageCoderFactory = value;
				if (this.messageCoderFactory != null)
				{
					writerStorage.MessageCoder = this.messageCoderFactory.newCoder(null);
				}
			}
			
		}

		virtual public AsyncCallManager AsyncCallManager
		{
			get
			{
				return this.asyncCallMgr;
			}
			
		}
		virtual public ConnectorFactory ConnectorFactory
		{
			get
			{
				return this.conFactory;
			}
			
		}
		virtual public AcceptorFactory AcceptorFactory
		{
			get
			{
				return this.acpFactory;
			}
			
		}
		
		private const string scheme = "bnmq";
		
		protected internal WriterStorage writerStorage = new WriterStorage();
		//protected internal ReaderStorage readerStorage = new ReaderStorage();		
		protected internal ConnectorFactory conFactory;
		protected internal AcceptorFactory acpFactory;
		
		protected internal Thread writerThread;
		protected internal Writer writerThreadBody;
		
		protected internal Thread readerThread;
		//protected internal Reader readerThreadBody;

		protected internal ITransportMessageCoderFactory messageCoderFactory;
		protected internal AsyncCallManager asyncCallMgr = new AsyncCallManager();
		
		public TransportFactory()
		{
            conFactory = new ConnectorFactory(writerStorage, this);
            acpFactory = new AcceptorFactory(writerStorage, this);
            startAsyncDispatchers();
		}
		
		public virtual ITransport getClientTransport(Uri addr)
		{
			return conFactory.getTransport(addr);
		}

        public virtual ITransport getServerTransport(Uri addr)
		{
			return acpFactory.getTransport(addr);
		}

        public virtual bool checkURISupport(Uri addr)
		{
			return addr.Scheme.ToUpper().Equals(scheme.ToUpper());
		}
		
		protected internal virtual void  startAsyncDispatchers()
		{
			writerThreadBody = new Writer(writerStorage);
			writerThread = new Thread(new ThreadStart(writerThreadBody.Run));
			writerThread.Name = "BNMQ-TCPWriter";
            writerThread.Start();
			/*readerThreadBody = new Reader(readerStorage);
			readerThread = new Thread(new ThreadStart(readerThreadBody.Run));
			readerThread.Name = "BNMQ-TCPReader";
			readerThread.Start();*/
		}

        public void close()
        {
            writerThreadBody.stop();
            //readerThreadBody.stop();
            writerStorage.close();
            //if(writerThread.IsAlive) - not supported on CF
                writerThread.Join();

            //readerThread.Join();
            //readerStorage.Finalize();

            conFactory.close();
            acpFactory.close();
            asyncCallMgr.stop();
        }
		
		~TransportFactory()
		{
            close();
		}
	}
}

File: BinaryNotesMQ/java/src/org/bn/mq/net/tcp/TransportFactory.java
/*
 * Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
 * 
 * Licensed under the LGPL, Version 2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.gnu.org/copyleft/lgpl.html
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * With any your questions welcome to my e-mail 
 * or blog at http://abdulla-a.blogspot.com.
 */

package org.bn.mq.net.tcp;

import java.io.IOException;

import java.net.URI;
import org.bn.mq.net.*;

public class TransportFactory implements ITransportFactory {
    private final String scheme = "bnmq";

    protected WriterStorage writerStorage = new WriterStorage();
    protected ReaderStorage readerStorage = new ReaderStorage();
    
    protected ConnectorFactory conFactory = new ConnectorFactory( writerStorage, readerStorage, this);
    protected AcceptorFactory acpFactory = new AcceptorFactory(writerStorage, readerStorage, this);

    protected Thread writerThread;
    protected Writer writerThreadBody;
    
    protected Thread readerThread;
    protected Reader readerThreadBody;
    protected ITransportMessageCoderFactory messageCoderFactory;
    protected AsyncCallManager asyncCallMgr = new AsyncCallManager();
    
    public TransportFactory () {
        startAsyncDispatchers();
    }        
    
    public void setTransportMessageCoderFactory(ITransportMessageCoderFactory coderFactory) {
        this.messageCoderFactory = coderFactory;
        if(this.messageCoderFactory!=null) {
            writerStorage.setMessageCoder(this.messageCoderFactory.newCoder(null));
        }
    }
    
    public ITransportMessageCoderFactory getTransportMessageCoderFactory() {
        return messageCoderFactory;
    }
    
    public AsyncCallManager getAsyncCallManager() {
        return this.asyncCallMgr;
    }
    
    public ITransport getClientTransport(URI addr) throws IOException {        
        return conFactory.getTransport(addr);
    }

    public ITransport getServerTransport(URI addr) throws IOException {        
        return acpFactory.getTransport(addr);
    }

    public boolean checkURISupport(URI addr) {
        return addr.getScheme().equalsIgnoreCase(scheme);
    }
    
    public ConnectorFactory getConnectorFactory() {
        return this.conFactory;
    }

    public AcceptorFactory getAcceptorFactory() {
        return this.acpFactory;
    }
    
    protected void startAsyncDispatchers() {
        writerThreadBody = new Writer(writerStorage);
        writerThread = new Thread(writerThreadBody);
        writerThread.setName("BNMQ-TCPWriter");
        writerThread.start();
        readerThreadBody = new Reader(readerStorage);
        readerThread = new Thread(readerThreadBody);
        readerThread.setName("BNMQ-TCPReader");
        readerThread.start();
    }
    
    public void close() {
        writerThreadBody.stop();
        readerThreadBody.stop();
        writerStorage.close();  
        if(writerThread.isAlive()) {
            try {
                writerThread.join();
            }
            catch(Exception ex) {
                ex = null;
            }
        }
        if(readerThread.isAlive()) {
            try {
                readerThread.join();    
            }
            catch(Exception ex) {
                ex = null;
            }            
        }
        
        readerStorage.close();
        conFactory.close();
        acpFactory.close();
        asyncCallMgr.stop();
    }
    
}
